metaci/api/views/test_testmethod_perf.py

```diff
@@ -0,0 +1,435 @@
+
+    def identical_tests_helper(self, count, method_name="GenericMethod", **fields):
+        t1 = TestResultFactory(
+            build_flow__tests_total=1, method__name=method_name, **fields
+        )
+        for i in range(count - 1):
+            TestResultFactory(**fields, build_flow__tests_total=1, method=t1.method)
+
+
+class TestTestMethodPerfRESTAPI(APITestCase, _TestingHelpers):
+    """Test the testmethodperf REST API"""
+
+    @classmethod
+    def setUpClass(cls):
+        cls.client, cls.user = cls.make_user_and_client()
+        super().setUpClass()
+
+    def setUp(self):
+        self.client.force_authenticate(self.user)
+        t1 = TestResultFactory(
+            method__name="Foo", duration=10, build_flow__tests_total=1
+        )
+        TestResultFactory(duration=2, build_flow__tests_total=1, method=t1.method)
+        TestResultFactory(method__name="Bar", duration=3, build_flow__tests_total=1)
+        TestResultFactory(method__name="Bar", duration=5, build_flow__tests_total=1)
+
+    def test_counting(self):
+        """Test counting of method invocations"""
+        objs = self.stats_test_helper("count")
+        self.assertEqual(self.find_by("method_name", objs, "Foo")["count"], 2)
+        self.assertEqual(self.find_by("method_name", objs, "Bar")["count"], 2)
+
+    def test_averaging(self):
+        """Test averaging of methods"""
+        objs = self.stats_test_helper("duration_average")
+
+        self.assertEqual(
+            self.find_by("method_name", objs, "Foo")["duration_average"], 6
+        )
+        self.assertEqual(
+            self.find_by("method_name", objs, "Bar")["duration_average"], 4
+        )
+
+    def test_all_included_fields(self):
+        includable_fields = [
+            "duration_average",
+            "duration_slow",
+            "duration_fast",
+            "duration_stddev",
+            "duration_coefficient_var",
+            "cpu_usage_average",
+            "cpu_usage_low",
+            "cpu_usage_high",
+            "count",
+            "failures",
+            "assertion_failures",
+            "DML_failures",
+            "Other_failures",
+            "success_percentage",
+        ]
+
+        def _test_fields(fields):
+            response = self.client.get(self.api_url(include_fields=fields))
+            self.assertEqual(response.status_code, 200)
+            rows = response.json()["results"]
+            for field in fields:
+                for row in rows:
+                    self.assertIn(field, row.keys())
+
+        random.seed("xyzzy")
+        for i in range(10):
+            field = random.sample(includable_fields, 1)
+            _test_fields(field)
+
+        for i in range(10):
+            field1, field2 = random.sample(includable_fields, 2)
+            _test_fields([field1, field2])
+
+        for i in range(10):
+            field1, field2, field3 = random.sample(includable_fields, 3)
+            _test_fields([field1, field2, field3])
```

Comment:
What's the rationale for doing this random sampling with a preset seed?

Reply:
The rationale for random sampling is that I don't want to make all of the combinations by hand. (roughly 30 lines of code)

The rational for presetting the seed is that if you don't, the tests might fail intermittently every 3 years for a mysterious reason and you won't be able to debug it.

I had a "commute-thought" today that another way to do it would be to have the seed be the current month (or year), so you would get "new data" every month but you would have a whole month to debug the issue before it would go away on its own.